airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams.py

```diff
@@ -0,0 +1,205 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from json import dumps, loads
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+from .common import URL_BASE, PageToken, SourceContext
+from .schemas import DisplayAdGroup, DisplayCampaign, DisplayProductAds, DisplayTargeting, JSModel, Profile, Types
+
+
+# Basic full refresh stream
+class AmazonAdsStream(HttpStream, ABC):
+    def __init__(self, config, *args, context: SourceContext = SourceContext(), **kwargs):
+        self.ctx = context
+        self._config = config
+        super().__init__(*args, **kwargs)
+
+    url_base = URL_BASE
+
+    @property
+    @abstractmethod
+    def model(self) -> JSModel:
+        """
+        Pydantic model to represent json schema
+        """
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_headers(self, *args, **kvargs) -> MutableMapping[str, Any]:
+        return {"Amazon-Advertising-API-ClientId": self._config.client_id}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an object representing single record in the response
+        """
+        yield from loads(response.text)
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+
+class Profiles(AmazonAdsStream):
+    """
+    This stream corresponds to Amazon Advertising API - Profiles
+    https://advertising.amazon.com/API/docs/en-us/reference/2/profiles#/Profiles
+    """
+
+    primary_key = "profileId"
+    model = Profile
+
+    def path(self, **kvargs) -> str:
+        return "v2/profiles"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        for record in super().parse_response(response, **kwargs):
+            profile_id_obj = self.model.parse_obj(record)
+            self.ctx.profiles.append(profile_id_obj)
+            yield record
+
+    def read_records(self, *args, **kvargs) -> Iterable[Mapping[str, Any]]:
+        if self.ctx.profiles:
+            yield from [profile.dict(exclude_unset=True) for profile in self.ctx.profiles]
+        else:
+            yield from super().read_records(*args, **kvargs)
+
+    def fill_context(self):
+        """
+        Fill profiles info for other streams in case of "profiles" stream havent been specified on catalog config
+        """
+        _ = [record for record in self.read_records(SyncMode.full_refresh)]
```

Comment:
using vcrpy help to break this hell of dependencies, please take a look at it.
We don't need to share common state, in most cases streams even don't need to have this state

Reply:
I liked vcrpy but it require additional rework/retest and Im short on time. Will change this on next iteration, thanks.